### PR TITLE
Fix: fix revision will change when add new trait with skiprevisionaffect to application

### LINKED ISSUE #2991: The newly added Trait with SkipRevisionAffect causes the Revision to change
**Describe the bug**
We are using context.appRevision as the label of the pod for the blue-green upgrade. The newly added Trait with SkipRevisionAffect causes the Revision to change. It will cause pod recreate. Is this a bug?

**To Reproduce**
definition yaml as follow:
```yaml
apiVersion: core.oam.dev/v1beta1
kind: ComponentDefinition
metadata:
  name: testapp
  namespace: vela-system
spec:
  workload:
    definition:
      apiVersion: apps/v1
      kind: Deployment
  schematic:
    cue:
      template: |
        output: {
          apiVersion: "apps/v1"
          kind:       "Deployment"
          metadata: {
            name: context.name
          }
          spec: {
            selector: matchLabels: {
              "app": context.name
            }
            template: {
              metadata: labels: {
                "app": context.name
                "harmonycloud.cn/bluegreen": context.appRevision
              }
              spec: {
                containers: [{
                  name: "nginx"
                  image: "nginx"
                  ports: [{
                    containerPort: 80
                    protocol: "TCP"
                  }]
                }]
              }
            }
          }
        }
---
apiVersion: core.oam.dev/v1beta1
kind: TraitDefinition
metadata:
  name: related-service
  namespace: vela-system
spec:
  podDisruptive: false
  skipRevisionAffect: true
  schematic:
    cue:
      template: |
        import "strings"
        import "strconv"

        parameter: {}

        outputs: service: {
            apiVersion: "v1"
            kind:       "Service"
            metadata: name: context.name
            spec: {
                selector: {
                    "app": context.name
                }
                ports: [
                    for _, v in {
                        for container in context.output.spec.template.spec.containers {
                            if container.ports != _|_ {
                                for i, p in container.ports {
                                    "\(strings.ToLower(p.protocol))-\(strconv.FormatInt(p.containerPort,10))": {
                                        name: strings.ToLower(p.protocol) + "-" + strconv.FormatInt(p.containerPort,10)
                                        port: p.containerPort
                                        targetPort: p.containerPort
                                    }
                                }
                            }
                        }
                } {v} ]
            }
        }
```
application yaml as follow:
```yaml
# app v1
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: testapp
spec:
  components:
  - type: testapp
    name: testapp
    properties: {}
---
# app v2
apiVersion: core.oam.dev/v1beta1
kind: Application
metadata:
  name: testapp
spec:
  components:
  - type: testapp
    name: testapp
    properties: {}
    traits:
    - type: related-service
      properties: {}
```
upgrade app from v1 to v2, it will create applicationrevision v2, and pod will recreate because pod label is context.appRevision

**Expected behavior**
upgrade app from v1 to v2 when just skipAffectRevision trait change, should not change application revision

**Screenshots**
First, app in version 1:
```
Every 1.0s: kubectl get po,app,applicationrevision,svc                                                                                 Sun Dec 19 13:43:17 2021

NAME                           READY   STATUS    RESTARTS   AGE
pod/testapp-84779fc575-jxg8n   1/1     Running   0          4s

NAME                               COMPONENT   TYPE	 PHASE     HEALTHY   STATUS   AGE
application.core.oam.dev/testapp   testapp     testapp   running   true               4s

NAME                                          AGE
applicationrevision.core.oam.dev/testapp-v1   4s

NAME                 TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE
service/kubernetes   ClusterIP   10.96.0.1    <none>        443/TCP   46d
```
upgrade app to v2
```
Every 1.0s: kubectl get po,app,applicationrevision,svc                                                                                 Sun Dec 19 13:43:54 2021

NAME                           READY   STATUS        RESTARTS   AGE
pod/testapp-66b88b6b68-xbpld   1/1     Running       0          4s
pod/testapp-84779fc575-jxg8n   0/1     Terminating   0          41s

NAME                               COMPONENT   TYPE	 PHASE     HEALTHY   STATUS   AGE
application.core.oam.dev/testapp   testapp     testapp   running   true               41s

NAME                                          AGE
applicationrevision.core.oam.dev/testapp-v1   41s
applicationrevision.core.oam.dev/testapp-v2   4s

NAME                 TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)   AGE
service/kubernetes   ClusterIP   10.96.0.1	<none>        443/TCP   46d
service/testapp      ClusterIP   10.96.99.211   <none>        80/TCP    4s
```

**KubeVela Version**
kubevela v1.1.13

**Cluster information**
```
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.2", GitCommit:"092fbfbf53427de67cac1e9fa54aaa09a28371d7", GitTreeState:"clean", BuildDate:"2021-06-16T12:59:11Z", GoVersion:"go1.16.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.2", GitCommit:"092fbfbf53427de67cac1e9fa54aaa09a28371d7", GitTreeState:"clean", BuildDate:"2021-06-16T12:53:14Z", GoVersion:"go1.16.5", Compiler:"gc", Platform:"linux/amd64"}
```

**Additional context**
Is this a bug? I still think like #2155 , skipAffectRevision traits should not be record in applicationrevision, and merge from current application when upgrade

### PR DESCRIPTION
**Pull request by chwetion**
### Description of your changes
Fixes #2991 

### Special notes for your reviewer
When I fixed the problem, I found that there are some dependencies between ApplicationRevision and Workflow, I think these should be two independent modules

### CODE REVIEW

**Comment by wangyikewxgm** [comment]
If modify a `skipAffectRevision` trait, is the changes will be updated in latest appRev?

**Comment by chwetion** [comment]
@wangyikewxgm It will have no change in appRev spec.traitDefinitions field, I also try it in 1.2.0-rc2

File: pkg/controller/core.oam.dev/v1alpha2/application/revision.go [hunk 1]
```diff
@@ -361,7 +361,11 @@ func (h *AppHandler) currentAppRevIsNew(ctx context.Context) (bool, bool, error)
 
 	// diff the latest revision first
 	if h.app.Status.LatestRevision.RevisionHash == h.currentRevHash && DeepEqualRevision(h.latestAppRev, h.currentAppRev) {
+		appSpec := h.currentAppRev.Spec.Application.Spec
+		traitDef := h.currentAppRev.Spec.TraitDefinitions
 		h.currentAppRev = h.latestAppRev.DeepCopy()
+		h.currentAppRev.Spec.Application.Spec = appSpec
+		h.currentAppRev.Spec.TraitDefinitions = traitDef
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/revision.go:368]: why we need to add these 4 lines here?

**Review by wangyikewxgm — APPROVED** [review]
discussed with @chwetion offline.  the behavior is meet expectation. LGTM

**Review by wonderflow — APPROVED** [review]

**Pull request merged**